src/lib.rs

```diff
@@ -22,29 +22,37 @@ However, CGS has a number of advantages:
 
 use std::marker::PhantomData;
 use std::{mem, ops};
-use std::sync::{Arc, Mutex, RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::sync::{Arc, Mutex, RwLock, RwLockReadGuard, RwLockWriteGuard, Weak};
 
 /// Reference counter type. It doesn't make sense to allocate too much bit for it in regular applications.
 // TODO: control by a cargo feature
 type RefCount = u16;
 
+struct PointerCounter {
+    strong: RefCount,
+    weak: RefCount,
+}
 /// Inner storage data that is locked by `RwLock`.
 struct StorageInner<T> {
     data: Vec<T>,
-    meta: Vec<RefCount>,
+    meta: Vec<PointerCounter>,
     free_list: Vec<usize>,
 }
 
 /// Pending reference counts updates.
 struct Pending {
-    add_ref: Vec<usize>,
-    sub_ref: Vec<usize>,
+    add_ref_strong: Vec<usize>,
+    sub_ref_strong: Vec<usize>,
+    add_ref_weak: Vec<usize>,
```

Comment:
Do we actually need to store the weak pointer count in the meta? It's not obvious to me, why we'd do that.